### ISSUE TITLE
Very slow sequential writes with NettyTcpClient

### ISSUE THREAD

**Issue by rstoyanchev** [reactor/reactor#551]
There is a noticeable lag when sending messages sequentially through the NettyTcpClient. The issue was reported as [Spring Boot ticket 4003](https://github.com/spring-projects/spring-boot/issues/4003) and results in very slow sending of sequential messages.

I've created a branch with a relatively focused example [ClientApp](https://github.com/rstoyanchev/sb-1.2-websocket/blob/client-app/src/main/java/org/krm1312/test/ClientApp.java) without the use of Spring. It requires a STOMP broker running (e.g. [RabbitMQ with the stomp plugin](https://www.rabbitmq.com/stomp.html)). Sample output:

````
18:17:37.519 [main] DEBUG org.krm1312.test.ClientApp - Sending message: 0
18:17:38.521 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] WRITE: 47B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 53 45 4e 44 0a 64 65 73 74 69 6e 61 74 69 6f 6e |SEND.destination|
|00000010| 3a 2f 74 6f 70 69 63 2f 66 6f 6f 0a 63 6f 6e 74 |:/topic/foo.cont|
|00000020| 65 6e 74 2d 6c 65 6e 67 74 68 3a 30 0a 0a 00    |ent-length:0... |
+--------+-------------------------------------------------+----------------+
18:17:38.522 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] FLUSH
18:17:38.523 [main] DEBUG org.krm1312.test.ClientApp - Sending message: 1
18:17:39.528 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] WRITE: 47B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 53 45 4e 44 0a 64 65 73 74 69 6e 61 74 69 6f 6e |SEND.destination|
|00000010| 3a 2f 74 6f 70 69 63 2f 66 6f 6f 0a 63 6f 6e 74 |:/topic/foo.cont|
|00000020| 65 6e 74 2d 6c 65 6e 67 74 68 3a 30 0a 0a 00    |ent-length:0... |
+--------+-------------------------------------------------+----------------+
18:17:39.528 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] FLUSH
18:17:39.529 [main] DEBUG org.krm1312.test.ClientApp - Sending message: 2
18:17:40.529 [reactor-tcp-io-1] DEBUG r.i.n.impl.netty.tcp.NettyTcpClient - [id: 0x8b679ca7, /127.0.0.1:56796 => /127.0.0.1:61613] WRITE: 47B
````

From some debugging it looks like writes are scheduled/queued with the Netty channel. However, each write takes a second to complete. In some cases (when running with Spring's WebSocket support) I've noticed even longer 10 second delays where the write happens when a heartbeat comes in. It's almost as if Netty doesn't notice there is a pending write until some other I/O event happens.

**Comment by smaldini** [comment]
Investigating, for now my log reports no issues so I wonder if it is JDK/system dependant

```

2015-09-28 13:10:00.481  INFO 13472 --- [eactor-tcp-io-1] o.s.m.s.s.StompBrokerRelayMessageHandler : BrokerAvailabilityEvent[available=true, StompBrokerRelay[127.0.0.1:61613]]
Sending msg: 0
[...]
Sending msg: 999
Sent 999 messages in 245.1 ms
```

**Comment by rstoyanchev** [comment]
I recall maybe once or twice not getting the issue. Perhaps something concurrency related. I'll also try different JDK versions.

**Comment by krm1312** [comment]
ProductName:	Mac OS X
ProductVersion:	10.10.5
BuildVersion:	14F27

java version "1.8.0_51"
Java(TM) SE Runtime Environment (build 1.8.0_51-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.51-b03, mixed mode)

I also tried on Linux, same issue.

java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)

Scientific Linux release 7.1 (Nitrogen)

**Comment by smaldini** [comment]
My bad , *I do actually reproduce the issue*, I was on the master branch... Investigating now with more brain :)

**Comment by smaldini** [comment]
If you can please try by forcing 2.0.6.BUILD-SNAPSHOT dependency like this

```xml

    <repositories>
        <repository>
            <id>spring snapshots</id>
            <url>http://repo.spring.io/snapshot</url>
        </repository>
    </repositories>


    <dependencies>
        <!--<dependency>-->
            <!--<groupId>org.springframework.boot</groupId>-->
            <!--<artifactId>spring-boot-starter-amqp</artifactId>-->
        <!--</dependency>-->

        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-websocket</artifactId>
        </dependency>

        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>17.0</version>
        </dependency>

        <!-- Uncomment for SB 1.3 -->
        <dependency>
            <groupId>io.projectreactor</groupId>
            <artifactId>reactor-core</artifactId>
            <version>2.0.6.BUILD-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>io.projectreactor</groupId>
            <artifactId>reactor-bus</artifactId>
            <version>2.0.6.BUILD-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>io.projectreactor</groupId>
            <artifactId>reactor-stream</artifactId>
            <version>2.0.6.BUILD-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>io.projectreactor</groupId>
            <artifactId>reactor-net</artifactId>
            <version>2.0.6.BUILD-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-all</artifactId>
            <version>4.0.31.Final</version>
        </dependency>

        <!-- Uncomment for SB 1.2.6 -->
        <!--<dependency>-->
            <!--<groupId>org.projectreactor</groupId>-->
            <!--<artifactId>reactor-net</artifactId>-->
        <!--</dependency>-->

    </dependencies>
```

**Comment by krm1312** [comment]
Looks good with the standalone ClientApp.

```
06:23:56.558 [main] INFO  org.krm1312.test.ClientApp - Sent 1000 in 470.4 ms
```

Thanks!

One thing I will point out, however, is full spring boot no longer starts with this version.  Fails with error:

```
Caused by: java.lang.NoSuchMethodError: reactor.io.net.impl.netty.NettyClientSocketOptions.eventLoopGroup(Lio/netty/channel/nio/NioEventLoopGroup;)Lreactor/io/net/impl/netty/NettyClientSocketOptions;
	at org.springframework.messaging.tcp.reactor.Reactor2TcpClient$1.apply(Reactor2TcpClient.java:111)
	at org.springframework.messaging.tcp.reactor.Reactor2TcpClient$1.apply(Reactor2TcpClient.java:104)
	at reactor.io.net.NetStreams.tcpClient(NetStreams.java:532)
	at org.springframework.messaging.tcp.reactor.Reactor2TcpClient.connect(Reactor2TcpClient.java:188)
	at org.springframework.messaging.simp.stomp.StompBrokerRelayMessageHandler.startInternal(StompBrokerRelayMessageHandler.java:405)
	at org.springframework.messaging.simp.broker.AbstractBrokerMessageHandler.start(AbstractBrokerMessageHandler.java:164)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:173)
	... 18 more
```

**Comment by smaldini** [comment]
Ay didn't realize that, but easy to fix, will coordinate with @rstoyanchev for the spring update.

**Comment by smaldini** [comment]
BTW in Client App, one possibility (not exposed in Spring yet) is to call writeWith only once with Streams.createWith() or any Publisher with more than one element you can read on demand. This will allow the request call to proceed in the Netty event loop already as the slowness penalty is mainly applied when writeWith is called outside the event loop thread.

**Comment by rstoyanchev** [comment]
Looks good .. also I applied an update for Spring Framework 4.2.2 for the NoSuchMethodError https://github.com/spring-projects/spring-framework/commit/c75206f9752ba532219085589817677a666f39bb.

**Comment by rstoyanchev** [comment]
Thanks!

**Comment by wilkinsona** [comment]
Thank you, @rstoyanchev and @smaldini.

@smaldini Can we please have a Reactor 2.0.6.RELEASE in time for Boot's 1.3.0.RC1 release which is currently scheduled for 7 October.

**Comment by smaldini** [comment]
should be in time, the other issue I'm looking at is #510

**Comment by smaldini** [comment]
@wilkinsona right the #510 is closed so virtually available asap for 2.0.6 RELEASE (before 7/10)

**Comment by wilkinsona** [comment]
@smaldini Lovely stuff. Thanks!
